pilot/pkg/bootstrap/server.go

```diff
@@ -184,10 +223,24 @@ func NewServer(args PilotArgs) (*Server, error) {
 	if err := s.initClusterRegistries(&args); err != nil {
 		return nil, fmt.Errorf("cluster registries: %v", err)
 	}
+
+	// Will run the sidecar injector in pilot - this precedes istiod, which does the same thing but with different
+	// config - certs are local. The code in main() hardcodes port 15017 and InjectionDirectory /etc/istio/inject
+	// and only operates if the directory exists - which is not set in either old or new installer. This code
+	// was used before sidecar injector was split to a separate deployment. Instead of attempting to use the old
+	// code, we'll use the istiod code which was tested more recently.
 	if err := s.initSidecarInjector(&args); err != nil {
 		return nil, fmt.Errorf("sidecar injector: %v", err)
 	}
 
+	if err := s.initIstiod(&args); err != nil {
```

Comment:
This doesn't make much sense to me. 

Why are we creating certs again, when we already initialize them in `initCertController`? 

Why is this "init istiod"? Is istiod a subset of pilot? That seems confusing to me, should probably just be `initGalley`, `initFoo`?

Reply:
initCertController is creating certs for other services ( via secrets ), in the old model where pilot is generating the DNS cert for inject/galley. 
It had some code to create pilot certs - but didn't work ( directory was RO).

I split this into components - initial method was just moved from istiod main.

Reply:
It seemed to work for me, maybe I did some workaround in the install though. Should we be reading the list of certs to create from meshconfig.certificates instead of hard coded list? can be a follow up

Reply:
The meshconfig.certificates was intended to provide certs for injector/galley, using secrets. We're not using that for Istiod itself. 

I changed the list a bit, take a look - it is now based on mesh.yaml discoveryAddress - will need few iterations to handle 'external' names, but the main point is if discoveryAddress is what pods will use when they're injected - and cert must match that.  I think we'll switch that to 
use .svc - i.e. istiod-canary.istio-system.svc:15012 

If control plane is external and we're just injecting - there is an env to override this.